builtin/plugins/dposv3/dpos.go

```diff
@@ -1336,6 +1390,24 @@ func UpdateDowntimeRecord(ctx contract.Context, validatorAddr loom.Address) erro
 		"down-blocks", statistic.RecentlyMissedBlocks&0xFFFF,
 	)
 
+	// if DPOSv3.3 enabled, jail a valdiator that have been offline for last 4 periods
+	if ctx.FeatureEnabled(loomchain.DPOSVersion3_3, false) {
+		state, err := loadState(ctx)
+		if err != nil {
+			return err
+		}
+		if statistic.RecentlyMissedBlocks&0xFFFF == state.Params.DowntimePeriod &&
```

Comment:
Instead of rewriting the bitstuffing, you could refactor `getDowntimeRecord` to accept a statistic. Then you'd do:

```go

if getDowntimeRecord(statistic) != nil {
     statistic.Jailed = true
}
```

Reply:
`getDowntimeRecord` returns the downtime record of a validator. So we still need to check if the validator is offline in the last 4 periods before jailing them. Anyway, refactored the bit bit-munging code, you can take a look.